qkeras/quantizers.py

```diff
@@ -440,6 +441,555 @@ def non_trainable_variables(self):
+        clipped_scaled_x = K.clip(scaled_x, self.int_repr_min, self.int_repr_max)
+        int_xq = _round_through(
+            clipped_scaled_x,
+            use_stochastic_rounding=self.use_stochastic_rounding,
+            precision=1.0,
+        )
+
+        return int_xq
+
+    def _get_alpha_scale(self, x):
+        """Get the minimum floating point scale that does not clip the max of x"""
+
+        axis = self._get_axis(x)
+
+        def alpha_scale_keep_negative():
+            """Get alpha scale when keeping negative values"""
+
+            return (K.max(tf.math.abs(x), axis=axis, keepdims=True) * 2) / self.int_repr_range
+
+        def alpha_scale_no_negative():
+            """Get alpha scale when dropping negative values"""
+
+            return K.max(x, axis=axis, keepdims=True) / self.int_repr_range
+
+        alpha_scale = tf.cond(
+            tf.equal(self.keep_negative, 1.0),
+            alpha_scale_keep_negative,
+            alpha_scale_no_negative,
+        )
+
+        return tf.math.maximum(alpha_scale, K.epsilon())
+
+    def _get_axis(self, x):
+        """Get axis for alpha scale computation"""
+
+        len_axis = tf.rank(x)
+        axis = tf.cond(
+            tf.not_equal(len_axis, 1),
+            lambda: _get_scaling_axis(self.scale_axis, len_axis),
+            lambda: tf.convert_to_tensor([0]),
+        )
+        return axis
+
+    def _po2_autoscale(self, x, alpha_scale):
+        """Get an approximation of the "best" po2 scale using least squares"""
+
+        # set alpha scale to a near power of two
+        alpha_scale = self._po2_round(alpha_scale)
+
+        def loop_body(_, alpha_scale, __):
+            """Loop body for least squares autoscaling"""
+
+            int_xq = self._get_quantized_integer(x, alpha_scale)
+            new_alpha_scale = _get_scale(
+                alpha="auto_po2",
+                x=x,
+                q=int_xq,
+                scale_axis=self.scale_axis,
+            )
+            return alpha_scale, new_alpha_scale, int_xq
+
+        def loop_cond(last_alpha_scale, alpha_scale, __):
+            """Loop condition for least squares autoscaling- stop when the scale converges"""
+
+            tensors_not_equal = tf.math.reduce_any(
+                tf.not_equal(last_alpha_scale, alpha_scale)
+            )
+            return tensors_not_equal
+
+        # Need a tensor of the same shape as alpha_scale that is not equal to alpha_scale
+        dummy_alpha_scale = -tf.ones_like(alpha_scale)
+
+        _, alpha_scale, int_xq = tf.while_loop(
+            loop_cond,
+            loop_body,
+            (dummy_alpha_scale, alpha_scale, x),
+            shape_invariants=(
+                alpha_scale.shape,
+                alpha_scale.shape,
+                tf.TensorShape(None),
+            ),
+            maximum_iterations=5,
+        )  # x and dummy_alpha_scale not used as inputs, just needed to determine shapes
+
+        return alpha_scale, int_xq
```

Comment:
It is a bit unintuitive that this function returns two variables. The function doc says `Get an approximation of the "best" po2 scale using least squares`, but we also return the integer representation of x with `return alpha_scale, int_xq`. Could make this function only return the alpha scale?

Reply:
Totally redid this- integrates much better with the multi-bit code